runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,216 @@
+  for _, m in pairs(token.modifiers) do
+    hl = modifiers_cache[m]
+    if hl then
+      highlight(buf, token, hl)
+    end
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/full`
+---
+function M.on_handle_full(err, response, ctx, _)
+  active_requests[ctx.bufnr] = false
+  vim.api.nvim_buf_clear_namespace(ctx.bufnr, ns, 0, -1)
+  -- if tick is has changed our response is outdated!
+  if err or last_tick[ctx.bufnr] ~= vim.api.nvim_buf_get_changedtick(ctx.bufnr) then
+    semantic_tokens[ctx.bufnr][ctx.client_id] = {}
+    return
+  end
+  local client = vim.lsp.get_client_by_id(ctx.client_id)
+  if not client then
+    return
+  end
+  local legend = client.server_capabilities.semanticTokensProvider.legend
+  local token_types = legend.tokenTypes
+  local token_modifiers = legend.tokenModifiers
+  local data = response.data
+  local ft = vim.api.nvim_buf_get_option(ctx.bufnr, "filetype")
+  local ft_token_cache = token_cache[ft]
+  local ft_modifiers_cache = modifiers_cache[ft]
+
+  local tokens = {}
+  local line, start_char = nil, 0
+  for i = 1, #data, 5 do
+    local delta_line = data[i]
+    line = line and line + delta_line or delta_line
+    local delta_start = data[i + 1]
+    start_char = delta_line == 0 and start_char + delta_start or delta_start
+
+    -- data[i+3] +1 because Lua tables are 1-indexed
+    local token_type = token_types[data[i + 3] + 1]
+    local modifiers = modifiers_from_number(data[i + 4], token_modifiers)
+
+    local token = {
+      line = line,
+      start_char = start_char,
+      length = data[i + 2],
+      type = token_type,
+      modifiers = modifiers
+    }
+    tokens[line + 1] = tokens[line + 1] or {}
+    table.insert(tokens[line + 1], token)
+
+    if token_type then
+      highlight_token(ctx.bufnr, token, ft_token_cache, ft_modifiers_cache)
+    end
+  end
+
+  if semantic_tokens[ctx.bufnr] then
+    semantic_tokens[ctx.bufnr][ctx.client_id] = tokens
+  else
+    semantic_tokens[ctx.bufnr] = {[ctx.client_id] = tokens}
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/refresh`
+---
+function M.on_refresh(err, _result, ctx, _config)
+  if not err then
+    M.refresh(ctx.bufnr)
+  end
+end
+
+---@private
+function M._save_tick(bufnr)
+  last_tick[bufnr] = vim.api.nvim_buf_get_changedtick(bufnr)
+  active_requests[bufnr] = true
+end
+
+--- Get currently active semantic tokens
+--
+-- @param bufnr number|nil
+-- @param client_id number|nil
+function M.get(bufnr, client_id)
+  if not bufnr then
+    return semantic_tokens
+  elseif not client_id then
+    return semantic_tokens[bufnr]
+  else
+    return semantic_tokens[bufnr][client_id]
+  end
+end
+
+--- Refresh the semantic tokens for the current buffer
+---
+--- It is recommended to trigger this using an autocmd or via keymap.
+---
+--- <pre>
+---   autocmd BufEnter,CursorHold,InsertLeave <buffer> lua require 'vim.lsp.semantic_tokens'.refresh()
+--- </pre>
+---
+--- @param bufnr number|nil
```

Comment:
Is there a bot to refresh the docs in lsp.txt or do I have to copy all relevant documentation?

Reply:
I found `gen_vimdoc.py`

Reply:
Ah, yes, sorry, forgot about that. Yes, that is run automatically.

Reply:
I still committed the relevant parts from this commit. Just to see that the docs get parsed correctly.